lib/perso.ml

```diff
@@ -1331,6 +1663,41 @@ let get_nb_marriage_witnesses_of_kind fam wk =
     (fun acc (_, w) -> if wk = w then acc + 1 else acc)
     0 witnesses
 
+let number_of_descendants_aux conf base env all_levels sl eval_int =
+  match get_env "level" env with
+  | Vint i -> (
+      match get_env "desc_level_table" env with
+      | Vdesclevtab t ->
+          let m = fst (Lazy.force t) in
+          let cnt =
+            Gwdb.Collection.fold
+              (fun cnt ip ->
+                if all_levels then
+                  if Gwdb.Marker.get m ip <= i then cnt + 1 else cnt
+                else if Gwdb.Marker.get m ip = i then cnt + 1
+                else cnt)
+              0 (Gwdb.ipers base)
+          in
+          VVstring (eval_int conf (if all_levels then cnt - 1 else cnt) sl)
+      | _ -> raise Not_found)
+  | _ -> raise Not_found
+
+let cous_paths_dates_aux conf base env p l1 l2 rev =
+  let _cousins_cnt, cousins_dates =
+    match (!cousins_t, !cousins_dates_t) with
+    | Some t, Some d_t -> (t, d_t)
+    | _, _ -> init_cousins_cnt conf base env p
+  in
+  match (int_of_string_opt l1, int_of_string_opt l2) with
+  | Some i, Some j -> (
+      let l =
+        if rev then List.rev cousins_dates.(i).(j) else cousins_dates.(i).(j)
```

Comment:
Here we use the `rev` parameter to choose if we want to maximum or minimum date.
But this assumes `cousins_dates` to always be sorted in increasing (or decreasing?) order, which may change one day causing a bug.
Why do we need to have cousins_dates.(i).(j) sorted?
if I understand correctly it is sorted because of the `List.sort_uniq` in `get_min_max_dates`.
If we use the fact that this list is sorted, we should have a comment explaining that it is important to have it sorted and why.

Also in the case of `rev = true`, we build a new list just to get the last element, if cousins_dates.(i).(j) is big this is a little wasteful

Reply:
Ok with those comments.
What is the "right" way to get the last element of the list?
Or mayby we can assume that the list is not sorted and we scan it for min or max. Would that be better?

Reply:
>  What is the "right" way to get the last element of the list?
A little verbose but I think something like:
```ocaml
let rec last = function
| [] -> failwith "empty"
| [x] -> x
| _x:: l -> last l
```
`list` is not the ideal data type when we need to work with elements others than the first ones.

>  Or mayby we can assume that the list is not sorted and we scan it for min or max. Would that be better?

yes but that would cost a little more, it is probably negligible.
Do we sort `cousins_dates.(i).(j)` for other reasons?